### PR TITLE
feat(ssa refactor): Implement first-class references

### LINKED ISSUE #1702: Add second-class references to Noir
### Problem

Noir has mutable variables, but they are strictly local: no mutation can be observed across function boundaries. This leads to code like the following:

```rs
fn foo() {
    let mut ctx = Context::new();
    ctx.foo.bar.my_vec = ctx.foo.bar.my_vec.push(4);
}
```

### Happy Case

If references were added, `BoundedVec::push` could take a mutable reference to the vector and the code could be simplified to the more natural:

```rs
fn foo() {
    let mut ctx = Context::new();
    ctx.foo.bar.my_vec.push(4);
}
```

However, adding first-class references comes with a host of problems. Namely, tracking the references as they are passed around everywhere, stored in structs, arrays, or returned from functions can become trouble-some. To alleviate this, we should only add second-class references to Noir. Second-class references will _not_ be allowed to be stored within structs or arrays, and can only be used essentially as a parameter-passing mode. This way our existing analysis for references will suffice while still gaining us some increased ergonomics.

### Alternatives Considered

_No response_

### Additional Context

_No response_

### Would you like to submit a PR for this Issue?

No

### Support Needs

_No response_

### PR DESCRIPTION
**Pull request by jfecher**
# Description

## Problem\*

Resolves #1702 - although I opted for first-class references instead of second-class references. These turned out to be easier/cleaner to add since adding second-class references would require some restrictions in how users are allowed to pass these around which are non-trivial to implement.

## Summary\*

Implements first-class references:
- A new type `&mut T` is now added representing mutable references.
  - Unlike rust, there are no lifetimes on references types in noir.
- New unary operators have been added:
  - Create a mutable reference via `&mut expr`
  - Dereference via `*expr`
- Dereferencing is also valid in the "lvalue" of an assignment.
  - In other words, the `*` in `*foo = bar;` is now allowed on the left hand side of the `=`.
  - This is required if `foo : &mut T`.
- An error is issued if:
  - Users use `&mut v` where `v` is an immutable variable or field of one.
  - Users use `&mut v[i]` where `v[i]` is an array access. Since arrays are immutable in the ssa refactor, this is somewhat more difficult to add so I opted to leave it as an error for now. Note that this is a user-facing error rather than a panic. The error message suggests storing the array element in another variable as a work around.

## Documentation

- [x] This PR requires documentation updates when merged.

  

  - [ ] I will submit a noir-lang/docs PR.

  

  - [x] I will request for and support Dev Rel's help in documenting this PR.
    - We should mention noir has mutable references.
    - Users can create mutable references via `&mut value` where `value` is either an existing (mutable) variable, or a literal value.
      - If it is an existing mutable variable, the reference will be aliased to the same variable
      - If it is to a literal value, a fresh reference will be created.
      - If it is to an immutable value, an error will be issued.
    - Mutable references have the type `&mut T` where `T` is the element type. For example, a mutable reference to a `Field` will have the type `&mut Field`.
    - Mutable references can be dereferenced to access their inner value: `*my_ref`. We can think of the unary `*` operation as having the type `fn(&mut T) -> T`.
    - Mutable references can be passed to functions to have functions mutate values visible to the caller.

  
  

## Additional Context

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

**Comment by jfecher** [comment]
This PR is passing again

File: crates/noirc_evaluator/src/ssa_refactor/ssa_gen/mod.rs [hunk 1]
```diff
@@ -161,14 +161,39 @@ impl<'a> FunctionContext<'a> {
...
+                        value::Value::Mutable(reference, _) => reference.into(),
+                    }
+                })
+            }
+            noirc_frontend::UnaryOp::Dereference => {
+                let typ = Self::convert_type(&unary.result_type);
+                rhs.map_both(typ, |rhs, element_type| {
+                    let rhs = rhs.eval(self);
+                    self.builder.insert_load(rhs, element_type).into()
+                })
```
Comment (jfecher) [inline crates/noirc_evaluator/src/ssa_refactor/ssa_gen/mod.rs:195]: update me

File: crates/noirc_frontend/src/hir/resolution/resolver.rs [hunk 2]
```diff
@@ -1010,37 +1024,38 @@ impl<'a> Resolver<'a> {
...
     fn resolve_pattern(&mut self, pattern: Pattern, definition: DefinitionKind) -> HirPattern {
-        self.resolve_pattern_mutable(pattern, None, definition)
+        self.resolve_pattern_mutable(pattern, false, None, definition)
     }
 
     fn resolve_pattern_mutable(
         &mut self,
         pattern: Pattern,
-        mutable: Option<Span>,
+        mutable: bool,
```
Comment (jfecher) [inline crates/noirc_frontend/src/hir/resolution/resolver.rs:1033]: revert this too

File: crates/noirc_frontend/src/hir_def/stmt.rs [hunk 3]
```diff
@@ -52,6 +52,7 @@ pub struct HirConstrainStatement(pub ExprId, pub FileId);
 pub enum HirPattern {
     Identifier(HirIdent),
     Mutable(Box<HirPattern>, Span),
+    MutableReference(Box<HirPattern>, Span),
```
Comment (jfecher) [inline crates/noirc_frontend/src/hir_def/stmt.rs:55]: remove

File: crates/noirc_frontend/src/monomorphization/printer.rs [hunk 4]
```diff
@@ -11,8 +11,8 @@ pub struct AstPrinter {
 
 impl AstPrinter {
     pub fn print_function(&mut self, function: &Function, f: &mut Formatter) -> std::fmt::Result {
-        let params = vecmap(&function.parameters, |(id, mutable, name, typ)| {
-            format!("{}{}$l{}: {}", if *mutable { "mut " } else { "" }, name, id.0, typ)
+        let params = vecmap(&function.parameters, |(id, mutability, name, typ)| {
+            format!("{}{}$l{}: {}", mutability, name, id.0, typ)
```
Comment (jfecher) [inline crates/noirc_frontend/src/monomorphization/printer.rs:15]: revert me

File: crates/noirc_frontend/src/monomorphization/printer.rs [hunk 5]
```diff
@@ -245,7 +245,12 @@ impl AstPrinter {
     ) -> Result<(), std::fmt::Error> {
         self.print_expr(&call.func, f)?;
         write!(f, "(")?;
-        self.print_comma_separated(&call.arguments, f)?;
+        for (i, elem) in call.arguments.iter().enumerate() {
+            self.print_expr(elem, f)?;
+            if i != call.arguments.len() - 1 {
+                write!(f, ", ")?;
+            }
+        }
```
Comment (jfecher) [inline crates/noirc_frontend/src/monomorphization/printer.rs:253]: revert me

File: crates/nargo_cli/tests/test_data_ssa_refactor/tuples/src/main.nr [hunk 6]
```diff
@@ -1,25 +1,25 @@
...
-    };
-    assert(u == x+1);
-    assert(v == x);
+    // let (u,v) = if x as u32 < 1 {
+    //     (x, x + 1)
+    // } else {
+    //     (x + 1, x)
+    // };
+    // assert(u == x+1);
+    // assert(v == x);
```
Comment (jfecher) [inline crates/nargo_cli/tests/test_data_ssa_refactor/tuples/src/main.nr:18]: revert

File: crates/noirc_evaluator/src/ssa_refactor/ssa_gen/context.rs [hunk 7]
```diff
@@ -553,7 +582,12 @@ impl<'a> FunctionContext<'a> {
                 }
             }
             (Tree::Leaf(lhs), Tree::Leaf(rhs)) => {
-                let (lhs, rhs) = (lhs.eval_reference(), rhs.eval(self));
+                let (lhs2, rhs2) = (lhs.clone().eval_reference(), rhs.clone().eval(self));
+
+                println!("Created store {lhs2} <- {rhs2} from {lhs:?} <- {rhs:?}");
+
+                let lhs = lhs2;
+                let rhs = rhs2;
```
Comment (jfecher) [inline crates/noirc_evaluator/src/ssa_refactor/ssa_gen/context.rs:590]: revert me

File: crates/noirc_evaluator/src/ssa_refactor/ssa_gen/context.rs [hunk 8]
```diff
@@ -535,9 +560,13 @@ impl<'a> FunctionContext<'a> {
                 self.assign_new_value(*array_lvalue, new_array.into());
             }
             LValue::MemberAccess { old_object, index, object_lvalue } => {
+                let old_object = old_object.map(|value| value.eval(self).into());
```
Comment (jfecher) [inline crates/noirc_evaluator/src/ssa_refactor/ssa_gen/context.rs:563]: verify removing me breaks the code

File: crates/noirc_evaluator/src/ssa_refactor/ssa_gen/context.rs [hunk 9]
```diff
@@ -473,11 +483,23 @@ impl<'a> FunctionContext<'a> {
...
+        let element_types = Self::convert_type(element_type);
+        values.map_both(element_types, |value, element_type| {
+            let reference = value.eval(self);
+            self.builder.insert_load(reference, element_type).into()
+        })
+    }
+
     /// Compile the given identifier as a reference - ie. avoid calling .eval()
-    fn ident_lvalue(&self, ident: &ast::Ident) -> Values {
+    fn ident_lvalue(&mut self, ident: &ast::Ident) -> Values {
```
Comment (jfecher) [inline crates/noirc_evaluator/src/ssa_refactor/ssa_gen/context.rs:502]: &self

File: crates/noirc_evaluator/src/ssa_refactor/ssa_gen/context.rs [hunk 10]
```diff
@@ -473,11 +483,23 @@ impl<'a> FunctionContext<'a> {
...
+            }
         }
     }
 
+    fn dereference(&mut self, values: &Values, element_type: &ast::Type) -> Values {
+        let element_types = Self::convert_type(element_type);
+        values.map_both(element_types, |value, element_type| {
+            let reference = value.eval(self);
+            self.builder.insert_load(reference, element_type).into()
+        })
```
Comment (jfecher) [inline crates/noirc_evaluator/src/ssa_refactor/ssa_gen/context.rs:498]: use this helper when evaluating de-referencing as well

**Comment by jfecher** [comment]
All of the outdated code has been removed

**Review by vezenovm — APPROVED** [review]

**Pull request merged**